karapace/protobuf/io.py

```diff
@@ -0,0 +1,248 @@
+import os
+
+ZERO_BYTE = b'\x00'
+
+logger = logging.getLogger(__name__)
+
+
+def calculate_class_name(name: str) -> str:
+    return "c_" + hashlib.md5(name.encode('utf-8')).hexdigest()
+
+
+class ProtobufDatumReader:
+    """Deserialize Protobuf-encoded data into a Python data structure."""
+
+    @staticmethod
+    def check_props(schema_one, schema_two, prop_list):
+        for prop in prop_list:
+            if getattr(schema_one, prop) != getattr(schema_two, prop):
+                return False
+        return True
+
+    @staticmethod
+    def match_schemas(writer_schema: ProtobufSchema, reader_schema: ProtobufSchema) -> bool:
+        # TODO (serge): schema comparison by fields required
+
+        if str(writer_schema) == str(reader_schema):
+            return True
+        return False
+
+    def __init__(self, writer_schema=None, reader_schema=None):
+        """
+    As defined in the Protobuf specification, we call the schema encoded
+    in the data the "writer's schema", and the schema expected by the
+    reader the "reader's schema".
+    """
+        self._writer_schema = writer_schema
+        self._reader_schema = reader_schema
+
+    # read/write properties
+    def set_writer_schema(self, writer_schema):
+        self._writer_schema = writer_schema
+
+    writer_schema = property(lambda self: self._writer_schema, set_writer_schema)
+
+    def set_reader_schema(self, reader_schema):
+        self._reader_schema = reader_schema
+
+    reader_schema = property(lambda self: self._reader_schema, set_reader_schema)
+
+    @staticmethod
+    def read_varint(bio: BytesIO) -> int:
+        """Read a variable-length integer.
+
+        :returns: Integer
+        """
+        varint = 0
+        read_bytes = 0
+
+        while True:
+            char = bio.read(1)
+            if len(char) == 0:
+                if read_bytes == 0:
+                    return 0
+                # raise EOFError('EOF while reading varint, value is %i so far' %
+                #               varint)
+
+            byte = ord(char)
+            varint += (byte & 0x7F) << (7 * read_bytes)
+
+            read_bytes += 1
+
+            if not byte & 0x80:
+                return varint
+
+    def read_indexes(self, bio: BytesIO):
+        size: int = self.read_varint(bio)
+        result = []
+        if size == 0:
+            result.append(0)
+            return result
+        i = 0
+        while i < size:
+            result.append(self.read_varint(bio))
+            i += 1
+
+    def read(self, bio: BytesIO):
+        if self.reader_schema is None:
+            self.reader_schema = self.writer_schema
+        return protobuf_to_dict(self.read_data(self.writer_schema, self.reader_schema, bio), True)
+
+    @staticmethod
+    def find_message_name(schema: ProtobufSchema, indexes: list) -> str:
+        result: list = []
+        dot: bool = False
+        types = schema.proto_file_element.types
+        for index in indexes:
+            if dot:
```

Comment:
looks like it might be easier to just unconditionally add a dot, and then remove the last dot at the end of the loop

Reply:
ok